[Regression Tests][hipCUB] Implementation of regression subtest for emulation (#425)
ROCm/hipCUB: CHANGELOG.md

## (Unreleased) hipCUB-x.x.x for ROCm 6.4.0

### Added

> * Added regression tests to `rtest.py`. Regression tests are a subset of tests that caused hardware problems for past emulation environments.

Reviewer: ```suggestion
* Added regression tests to `rtest.py`. These tests recreate scenarios that have caused hardware problems in past emulation environments. Use `python rtest.py [--emulation|-e|--test|-t]=regression` to run these tests.
```

>   * Can be run with `python rtest.py [--emulation|-e|--test|-t]=regression`

Reviewer: If you combine with the point above, you can remove this line.

Author: To keep things consistent with other changelogs for other primitive library, I think we should keep it as two lines.

* Added `ForEach`, `ForEachN`, `ForEachCopy`, `ForEachCopyN` and `Bulk` functions to have parity with CUB.
* Added the `hipcub::CubVector` type for CUB parity.

> * Added `--emulation` option for `rtest.py`

Reviewer: This is redundant since it was covered in the first point. It can be removed.

> * Unit tests can be ran with `[--emulation|-e|--test|-t]=<test_name>`

Reviewer: Which part of this was added? The --emulation option or something else? Or is it the script itself? In which case we can rewrite the first point about regression tests.

Author: The --emulation option

### Changed

> * Edited smoke test options, which runs a subset of the unit tests and ensuring that less than 2gb of VRAM will be used

Reviewer: Is it that the smoke tests have been changed to use less than 2GB of VRAM? Or is it that the smoke tests are now run in different way? That needs to be clarifued.

  * Smoke tests can be ran with `[--emulation|-e|--test|-t]=smoke`

> * `--test|-t` is no longer a required flag for `rtest.py`. Instead, the user can use either `--emulation|-e` or `--test|-t`, but not both.

Reviewer: I would put the change to rtest.py first.

```suggestion
### Changed
* The `rtest.py` options have changed.  `rtest.py` is now run with at least either `--test|-t` or `--emulation|-e`, but not both options.
```

### Added
* Added regression tests to `rtest.py`. These tests recreate scenarios that have caused hardware problems in past emulation environments. Use `python rtest.py [--emulation|-e|--test|-t]=regression` to run these tests.
* Added `ForEach`, `ForEachN`, `ForEachCopy`, `ForEachCopyN` and `Bulk` functions to have parity with CUB.
* Added the `hipcub::CubVector` type for CUB parity.
* Added `--emulation` option for `rtest.py`

>   * Unit tests can be ran with `[--emulation|-e|--test|-t]=<test_name>`

Reviewer: ```suggestion
  * Unit tests can be run with `[--emulation|-e|--test|-t]=<test_name>`
```

### Changed

> * Changed subset of tests to be ran for smoke tests such that it will complete in faster run-time and to never exceed 2GB of vram usage. Use `python rtest.py [--emulation|-e|--test|-t]=smoke` to run these tests.

Reviewer: ```suggestion
* Changed the subset of tests that are run for smoke tests such that the smoke test will complete with faster run-time and to never exceed 2GB of vram usage. Use `python rtest.py [--emulation|-e|--test|-t]=smoke` to run these tests.
```
